### ISSUE TITLE
pull/build of Docker image fails with error

### ISSUE THREAD

**Issue by jgbaum** [apptainer/singularity#3512]
## Version of Singularity:

3.1.1-1.el7.centos

### Expected behavior

Image should be pulled without error.

### Actual behavior

While pulling the image from a docker registry, the following error is encountered (URL replaced for security):

```
Writing manifest to image destination
Storing signatures
FATAL:   Unable to pull docker://REGISTRY_URL:PORT/bioinformatics/cluster-containers/torque: packer failed to pack: While unpacking tmpfs: unpack: error extracting layer: link /scratch/sbuild-986557175/fs/usr/share/terminfo/a/ansi80x25 /scratch/sbuild-986557175/fs/usr/share/terminfo/c/cons25: operation not permitted
```

### Steps to reproduce behavior

Here is the 'pull' command being run, as well as the resulting output:

```
$ singularity pull --docker-login docker://REGISTRY_URL:PORT/bioinformatics/cluster-containers/torque
WARNING: Authentication token file not found : Only pulls of public images will succeed
Enter Docker Username: jgbaum
Enter Docker Password: 
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:8ba884070f611d31cb2c42eddb691319dc9facf5e0ec67672fcfa135181ab3df
 71.91 MiB / 71.91 MiB [====================================================] 1s
Copying blob sha256:624932f6c2b97f2f92691dee730fb5fe177c9e4bf0c4cb981ed89521a7714587
 63.40 MiB / 63.40 MiB [====================================================] 1s
Copying blob sha256:cf4579f64bba88739ccc4f546a68b91d8015cc6167ff2fe256324a1bf9ac1554
 51.66 MiB / 51.66 MiB [====================================================] 1s
Copying blob sha256:d3fe1df916983a94266bcc727d11996d66369508e148883d3186c4ae20992785
 47.88 MiB / 47.88 MiB [====================================================] 1s
Copying blob sha256:0988771160c426823b6c5326ca0534771ad32d74fb0f05cebe5c09896e11cfed
 52.86 MiB / 52.86 MiB [====================================================] 1s
Copying blob sha256:dd55d9de0aac4dcdfff9599a992a4ef53574a918c7b45492f8260428c962e7cf
 82.45 MiB / 82.45 MiB [====================================================] 1s
Copying blob sha256:f32ab924acedf9f07ea3edc0aa989a45b5ae76abe86ff318da9b79b031381a87
 34.12 MiB / 34.12 MiB [====================================================] 1s
Copying blob sha256:817ef4c9a24d3a9ab8501d577ace1677ba5391d85f2dbfb1ce803e9f31fe035d
 34.89 MiB / 34.89 MiB [====================================================] 1s
Copying blob sha256:d9aab661c1e2ab58bdb4f7d41f3c30e5bb8ace3538f86ed71f1d3b62eed94506
 41.91 MiB / 41.91 MiB [====================================================] 1s
Copying blob sha256:fb0b5dcac520ac7e5d577dc15ab4687cac7fd3d1df1429047ec24fdb7e6b33e5
 5.00 MiB / 5.00 MiB [======================================================] 0s
Copying blob sha256:54f9193cc1124def9c2d551ca139d883a6f5aab29821fbe43678d368b7dd8d8a
 441.03 KiB / 441.03 KiB [==================================================] 0s
Copying blob sha256:5955fcb9f164f63ae8366c74456848d37d110a602647a3de103dbf328ba37b28
 40.16 MiB / 40.16 MiB [====================================================] 1s
Copying blob sha256:7233e6fd579558d3a8db758a3594d00414f40acec444b1ecf66f603941350a25
 12.76 MiB / 12.76 MiB [====================================================] 0s
Copying config sha256:07170f8cb49813b29b4e633e8e3fc9b9febb53061fb124cd12424d508cff3b5f
 3.43 KiB / 3.43 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
FATAL:   Unable to pull docker://REGISTRY_URL:PORT/bioinformatics/cluster-containers/torque: packer failed to pack: While unpacking tmpfs: unpack: error extracting layer: link /scratch/sbuild-546622671/fs/usr/share/terminfo/a/ansi80x25 /scratch/sbuild-546622671/fs/usr/share/terminfo/c/cons25: operation not permitted
```

Trying again, I get the same error:

```
$ singularity pull --docker-login docker://REGISTRY_URL:PORT/bioinformatics/cluster-containers/torque
WARNING: Authentication token file not found : Only pulls of public images will succeed
Enter Docker Username: jgbaum
Enter Docker Password: 
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:8ba884070f611d31cb2c42eddb691319dc9facf5e0ec67672fcfa135181ab3df
Skipping fetch of repeat blob sha256:624932f6c2b97f2f92691dee730fb5fe177c9e4bf0c4cb981ed89521a7714587
Skipping fetch of repeat blob sha256:cf4579f64bba88739ccc4f546a68b91d8015cc6167ff2fe256324a1bf9ac1554
Skipping fetch of repeat blob sha256:d3fe1df916983a94266bcc727d11996d66369508e148883d3186c4ae20992785
Skipping fetch of repeat blob sha256:0988771160c426823b6c5326ca0534771ad32d74fb0f05cebe5c09896e11cfed
Skipping fetch of repeat blob sha256:dd55d9de0aac4dcdfff9599a992a4ef53574a918c7b45492f8260428c962e7cf
Skipping fetch of repeat blob sha256:f32ab924acedf9f07ea3edc0aa989a45b5ae76abe86ff318da9b79b031381a87
Skipping fetch of repeat blob sha256:817ef4c9a24d3a9ab8501d577ace1677ba5391d85f2dbfb1ce803e9f31fe035d
Skipping fetch of repeat blob sha256:d9aab661c1e2ab58bdb4f7d41f3c30e5bb8ace3538f86ed71f1d3b62eed94506
Skipping fetch of repeat blob sha256:fb0b5dcac520ac7e5d577dc15ab4687cac7fd3d1df1429047ec24fdb7e6b33e5
Skipping fetch of repeat blob sha256:54f9193cc1124def9c2d551ca139d883a6f5aab29821fbe43678d368b7dd8d8a
Skipping fetch of repeat blob sha256:5955fcb9f164f63ae8366c74456848d37d110a602647a3de103dbf328ba37b28
Skipping fetch of repeat blob sha256:7233e6fd579558d3a8db758a3594d00414f40acec444b1ecf66f603941350a25
Copying config sha256:07170f8cb49813b29b4e633e8e3fc9b9febb53061fb124cd12424d508cff3b5f
 3.43 KiB / 3.43 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
FATAL:   Unable to pull docker://REGISTRY_URL:PORT/bioinformatics/cluster-containers/torque: packer failed to pack: While unpacking tmpfs: unpack: error extracting layer: link /scratch/sbuild-986557175/fs/usr/share/terminfo/a/ansi80x25 /scratch/sbuild-986557175/fs/usr/share/terminfo/c/cons25: operation not permitted
```
Looks potentially related to #2210 ?

**Comment by jgbaum** [comment]
A bit more information on this.   I ran into a similar issue by saving the docker image and using the docker_archive prefix to import:

```
$ singularity build torque.sif docker-archive://torque.dimg
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:d69483a6face4499acb974449d1303591fcbb5cdce5420f36f8a6607bda11854
 199.78 MiB / 199.78 MiB [=================================================] 27s
Copying blob sha256:fd5ec89dcf8c6417263780d0f01a08a2893da70f1d2a94120d07e8d9183caf9c
 181.61 MiB / 181.61 MiB [=================================================] 22s
Copying blob sha256:082f9b062500f165c256604cb81e56d03eb958f80711987bef2967c89b545a04
 215.31 MiB / 215.31 MiB [=================================================] 19s
Copying blob sha256:fb0da9faa4f24ba1842a4e1ab861be99e6efe1ed1c2556c01742e485eff26736
 138.44 MiB / 138.44 MiB [=================================================] 15s
Copying blob sha256:eda979b694da68a4cd491e829e5131506f465ed7bac08046347b48790042e554
 138.35 MiB / 138.35 MiB [=================================================] 17s
Copying blob sha256:55fd64984837cec96ba63c927a28a558f9c0a4a52b66d2c1b3044b20237b88c5
 251.87 MiB / 251.87 MiB [=================================================] 28s
Copying blob sha256:5dd762519fa13d2b68a0f1ef53f0e939f208e2ecc7e192cf0d7e0b7c15b519aa
 101.53 MiB / 101.53 MiB [=================================================] 13s
Copying blob sha256:e71aa4833379d50875100e67a2e571c58ce9d1595fa084ebb349311faf76f957
 105.08 MiB / 105.08 MiB [=================================================] 14s
Copying blob sha256:ac9118e07765522726754ce858f049fef7719f1fa66ed773b5af48341f417566
 60.82 MiB / 60.82 MiB [====================================================] 9s
Copying blob sha256:376030a7121be0703ad0711942dac911e9b3b8de7ed6a0db63d3de64108481f7
 21.30 MiB / 21.30 MiB [====================================================] 2s
Copying blob sha256:dd7fcb46b4901ad07a9516675e475a9ef4b6e6fa93e18652a15bf1f3ae79bf22
 2.91 MiB / 2.91 MiB [======================================================] 1s
Copying blob sha256:a34022966fe1ee4550ca215267f2a6dcf0ad3f9ab4529cd2f1c7bbc53a4cf050
 151.40 MiB / 151.40 MiB [=================================================] 17s
Copying blob sha256:53b6d57f9959dfef9ee2abb4e90695c49a1372550bd50a411cde11355c3c57e3
 36.71 MiB / 36.71 MiB [====================================================] 5s
Copying config sha256:07170f8cb49813b29b4e633e8e3fc9b9febb53061fb124cd12424d508cff3b5f
 3.43 KiB / 3.43 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
FATAL:   While performing build: packer failed to pack: While unpacking tmpfs: unpack: error extracting layer: link /scratch/sbuild-343035443/fs/usr/share/terminfo/a/ansi80x25 /scratch/sbuild-343035443/fs/usr/share/terminfo/c/cons25: operation not permitted
```

**Comment by jgbaum** [comment]
A bit more information...

The issue seems to be specific for this image (based off of [centos:7](https://hub.docker.com/_/centos)), as I'm able to successfully pull/build an image based off of [perl:5.28](https://hub.docker.com/_/perl), which itself is based off of Debian Stretch.

Any particular reason centos-based images would result in issues like this?

**Comment by jgbaum** [comment]
Confirming that if I pull centos:7 directly from Dockerhub, I run into the same issue:

```
$ singularity build c7.sif docker://centos:7
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:8ba884070f611d31cb2c42eddb691319dc9facf5e0ec67672fcfa135181ab3df
Copying config sha256:b9a1b1f0b2baaec83946a26d7045e4028f11eccc8b0e5b3514568e56a391beb2
 1.05 KiB / 1.05 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
FATAL:   While performing build: packer failed to pack: While unpacking tmpfs: unpack: error extracting layer: link /scratch/sbuild-704886214/fs/usr/share/terminfo/a/ansi80x25 /scratch/sbuild-704886214/fs/usr/share/terminfo/c/cons25: operation not permitted

```

**Comment by jgbaum** [comment]
On a different machine (desktop), running the same singularity version, I can build the image without issue.  I suspect it's something specific to the cluster architecture.  Any ideas would be much appreciated.

**Comment by jgbaum** [comment]
A few more bits of information:
- By becoming root on the cluster, I'm able to build successfully.
- By changing the --tmpdir to something other than the default, which is a beegfs filesystem, I'm able to build successfully.

The odd thing is that there are only certain images that are problematic with non-root access and the default temp dir.  I'll work with our cluster systems administrator to determine if there are settings that can be adjusted to make this work consistently.  In the meantime, any ideas are much appreciated.

**Comment by jgbaum** [comment]
Upon searching for beegfs in the closed issues, this appears to be related to #1476.  We'll find a workaround.
